external-crates/move/crates/move-package-alt/DESIGN.md

```diff
@@ -642,13 +633,68 @@ published-at = "TODO: replace this with the published address of your package"
 fixing it when they review the commit. We can also output a loud warning from the command line!
 
+## Ephemeral publish
+
+A common operation that we would like to support is to make an ephemeral publication of a package
+and its dependencies for testing (for example on localnet). To support this, we will add an
+ephemeral publish command that works like a normal publish except that the publication addresses for
+packages are read from and written to a separate file.
+
+The format of an ephemeral publication file is as follows:
+
+```toml
+# generated by move
+# this file contains metadata from ephemeral publications
+# this file should not be committed to source control
+
+build-env = "mainnet"
+chain-id = "localnet chain ID"
+
+[published.example]
+published-at = "0x000000000000000000000000000000000000000000000000000000000000cccc"
+original-id = "0x000000000000000000000000000000000000000000000000000000000000cc00"
+upgrade-cap = "0x000000000000000000000000000000000000000000000000000000000011cc00"
+
+[published.Foo_0]
+published-at = "0x0000000000000000000000000000000000000000000000000000000000001234"
+original-id = "0x0000000000000000000000000000000000000000000000000000000000005678"
+upgrade-cap = "0x000000000000000000000000000000000000000000000000000000000022cc00"
+```
+
+We will add a new command `sui client test-publish <pubfile>` which will publish build the root
+package for one environment and then publish it to a chain using the addresses from the ephemeral
+publication file.
+
+`sui client test-publish <pubfile> --build-env <env>` will work as follows:
+
+ - if `<pubfile>` is omitted, it defaults to `Pub.<env-name>.toml` where `<env-name>` is the name of
+   the active environment (irrespective of the environments defined in the manifest). This file
+   contains the addresses to use for publication
+
+ - if `--build-env <env>` is omitted, it defaults to the `build-env` name of the file at `<file>`;
+   if that is missing then we error. The build environment determines how we resolve dependencies
+   for the package
+
+ - if `--build-env <env>` is present and different from the `build-env` in the file, we
+   fail[^test-pub-mvp].
+
+ - if `<file>` exists and `chain-id` doesn't match, we fail[^test-pub-mvp].
+
+We then load the package graph from the lockfile for `build-env` (repinning if necessary). We check
+that all of the dependencies besides the root package are present in `<file>`[^test-pub-mvp]. We then publish the
+root package to the active environment using the addresses from `<file>` instead of their
+published-at addresses. Finally, we update the entry for the root package in `<file>`.
+
+[^test-pub-mvp]: We have a post-mvp proposal to automatically publish dependencies instead of failing if they
+    are missing; see "Recursive test-publish; test-upgrade"
+
 ## Post-MVP operations
 
-### Fixup lockfiles
+### Recursive test-publish; test-upgrade
```

Comment:
So today, folks could build things up iteratively by passing the same `<pubfile>` to each package in turn, right? I like that.